Introduce slice_inverse() op

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* #116308
* #117057
* __->__ #117041

Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)

@albanD / @soulitzer / @bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

jbschlosser committed:
Introduce slice_inverse() op

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2562,6 +2562,17 @@ Tensor slice(
+    const Tensor& self,
+    const Tensor& base,
+    int64_t dim,
+    c10::optional<SymInt> start,
```
These should be const ref especially since they aren't used. You may want to mark them as unused by commenting out the variable names. 
```suggestion
    c10::optional<SymInt> /*start*/,
```

jbschlosser replied:
Thanks for the review! I don't think I change to const ref here because the signature has to match the one in the generated header, but I can comment out the unused names.

Skylion007 commented on aten/src/ATen/FunctionalInverses.cpp:
```diff
@@ -428,12 +428,21 @@ Tensor FunctionalInverses::narrow_copy_inverse(const at::Tensor & base, const at
+    if (inverse_return_mode == InverseReturnMode::NeverView) {
+      return at::slice_copy_symint(mutated_view, dim, start, end, step);
+    } else {
+      return mutated_view.slice_symint(dim, start, end, step);
```
```suggestion
      return mutated_view.slice_symint(dim, std::move(start), std::move(end), std::move(step));
```
I would check slice_copy_symint call above too. c10::optional SymInt is not trivially copyable and should be moved.

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5361,14 +5361,21 @@
     CompositeExplicitAutograd: slice_backward
   autogen: slice_backward.out
 
+- func: slice_inverse(Tensor(a) self, Tensor src, int dim=0, SymInt? start=None, SymInt? end=None, SymInt step=1) -> Tensor(a)
```
Could you add a note here, describing the situation + exactly when this op gets used? I **think** two invariants we want to uphold are:

(1) This op shouldn't really show up in eager mode (so e.g. XLA shouldn't have to implement it)

(2) This op shouldn't show up in a PT2 graph (so a PT2 backend shouldn't have to implement it)

And the only case where this really shows up is during fakeification; meaning that a **subclass** might have to implement it (but only if we have a subclass that is an input to the PT2 graph, and is a view of some other tensor)

jbschlosser replied:
good call, added the note

bdhirsh commented on tools/autograd/derivatives.yaml:
```diff
@@ -1499,8 +1499,8 @@
 
-- name: slice_scatter(Tensor self, Tensor src, int dim=0, SymInt? start=None, SymInt? end=None, SymInt step=1) -> Tensor
-  self: slice_scatter_symint(grad, zeros_like(src), dim, start, end, step)
+- name: slice_inverse(Tensor(a) self, Tensor src, int dim=0, SymInt? start=None, SymInt? end=None, SymInt step=1) -> Tensor(a)
```
nit: maybe add a comment here saying that `slice_scatter` inherits its derivative formula from `slice_inverse`, since as far as autograd the two ops are the same, but have different aliasing behavior in the forward.

jbschlosser replied:
added!

bdhirsh commented on torchgen/model.py:
```diff
@@ -1616,6 +1617,9 @@ def strip_ret_annotation(r: Return) -> Return:
         if strip_view_copy_name and base_name.endswith("_copy"):
             base_name = base_name.replace("_copy", "")
 
+        if swap_scatter_inverse:
```
nit: I think that going from view_scatter-op to view_inverse-op is morally the same as going from view_copy to view op. So I don't think you need a new flag: you can just do:
```
        if strip_view_copy_name:
                if base_name.endswith("_copy"):
                    base_name = base_name.replace("_copy", "")
                elif base_name.endswith("_scatter")
                    base_name = base_name.replace("_scatter, "_inverse")
```

jbschlosser replied:
nice, makes it cleaner :)

bdhirsh commented on torchgen/model.py:
```diff
@@ -2603,9 +2607,9 @@ def __post_init__(self) -> None:
                 " See Note [view_copy NativeFunctions] for details."
             )
         else:
-            assert self.view_copy.func.name.name.base.endswith("_copy")
```
Nit: keep this assert, and assert that it ends with either `_copy` or `_scatter`?

jbschlosser replied:
sure that works!

bdhirsh commented on torchgen/model.py:
```diff
@@ -2659,6 +2663,10 @@ def gets_generated_view_copy(f: NativeFunction) -> bool:
     if "inplace_view" in f.tags:
         return False
+    # Assume ops ending in _inverse have manually-defined copy variants
+    # (e.g. slice_inverse() has the copy variant slice_scatter())
```
It might be helpful context to mention here why we don't just auto-gen the `slice_scatter` op (since we auto-gen every other view_copy op)

Short answer.. we probably could? (it's just cloning the base and then doing a slice into the cloned tensor). But the codegen is going to be slightly different, and making the codegen more complicated is probably more trouble than its worth compared to hand-writing ~5 kernels

bdhirsh commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -272,12 +275,16 @@ def unpacked_name(arg_name: str) -> str:
+    if g.view_copy is not None:
+        return ViewInverseSignature(g).name()
+    else:
+        # NB: some view ops don't have associated group data (e.g. chunk, narrow)
```
hmmm, why do we have the branching in this function?

In my head, we have total control over what the name of the view inverse function is (since the function here isn't even an ATen op, it's a shim that we implement ourselves in `FunctionalInverses.cpp`. So we should just name them whatever makes the codegen easiest to reason about.

jbschlosser replied:
The branching is because for *reasons* we don't have `NativeFunctionsViewGroup` info for `narrow()` / `chunk()` at least. But you're right that we can do whatever here as long as it matches what's done over in the functionalization codegen. If you're okay with me changing that codegen, I can probably remove the need to pass this info down, simplifying the codegen changes in this PR.

bdhirsh replied:
yep, changing the kernel names we codegen seems fine if it reduces complexity!

bdhirsh reviewed (commented):
Mostly left minor codegen nits, looks good!

We should add some tests for the new inverse ops. Probably ok if you want to wait til the next PR, and add tests that run view/view_inverse logic via autograd?

jbschlosser commented:
> We should add some tests for the new inverse ops. Probably ok if you want to wait til the next PR, and add tests that run view/view_inverse logic via autograd?

Note that I've already added tests that run the view / view_inverse replay logic in [`test_autograd.py`](https://github.com/pytorch/pytorch/blob/b6962208b85d35634294f6c620f6daf732f323b3/test/test_autograd.py#L8655) and an `OpInfo`-based set of tests in [`test_ops.py`](https://github.com/pytorch/pytorch/blob/b6962208b85d35634294f6c620f6daf732f323b3/test/test_ops.py#L1532), so the logic is exercised through those.

That said, I think the missing piece is to add an OpInfo entry directly for `slice_inverse()`, if we think that's worth it.

larryliu0820 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5361,14 +5361,21 @@
     CompositeExplicitAutogradNonFunctional: slice_scatter
   autogen: slice_scatter.out
-  tags: core
+  tags: [core, view_copy]
```
I don't have enough context on the whole PR but this looks fine on ExecuTorch codegen. I don't think we are consuming `view_copy` tag anyways.

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

bdhirsh commented:
I'm not too opinionated on whether or not its worth adding an OpInfo for that op. LGTM!

Skylion007 commented on aten/src/ATen/FunctionalInverses.cpp:
```diff
@@ -439,9 +439,9 @@ Tensor FunctionalInverses::slice_scatter_inverse(const at::Tensor & base, const
     auto output = mutated_view.slice_symint(dim, start, end, step);
     if (inverse_return_mode == InverseReturnMode::NeverView) {
-      return at::slice_copy_symint(mutated_view, dim, start, end, step);
+      return at::slice_copy_symint(mutated_view, dim, std::move(start), std::move(end), step);
```
```suggestion
      return at::slice_copy_symint(mutated_view, dim, std::move(start), std::move(end), std::move(step));
```

Skylion007 commented on aten/src/ATen/FunctionalInverses.cpp:
```diff
@@ -439,9 +439,9 @@ Tensor FunctionalInverses::slice_scatter_inverse(const at::Tensor & base, const
+      return at::slice_copy_symint(mutated_view, dim, std::move(start), std::move(end), step);
     } else {
-      return mutated_view.slice_symint(dim, start, end, step);
+      return mutated_view.slice_symint(dim, std::move(start), std::move(end), step);
```
```suggestion 
      return mutated_view.slice_symint(dim, std::move(start), std::move(end), std::move(step));
```

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

vadimkantorov commented:
Would also be nice to re-instate Python's op for torch.slice :) 
- https://github.com/pytorch/pytorch/issues/41625

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5361,14 +5361,29 @@
+# slice.Tensor, split_with_sizes, et. al.). Currently, these are only used during fake-ification
+# of PT2 graph input subclass instances that are views. This means:
+# * This op shouldn't really show up in eager mode (so e.g. XLA shouldn't have to implement it)
+# * This op shouldn't show up in a PT2 graph (so a PT2 backend shouldn't have to implement it)
```
don't compositeexplicit ops show in the graph?

jbschlosser replied:
in general yes they are, but we expect this particular op to only be invoked during fake-ification of an input subclass tensor instance, and op calls made during fake-ification aren't traced into the graph

lezcano replied:
I had completely missed that sentence. Ooops

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -1503,6 +1503,11 @@
+- name: slice_inverse(Tensor(a) self, Tensor src, int dim=0, SymInt? start=None, SymInt? end=None, SymInt step=1) -> Tensor(a)
+  self: grad.slice_symint(dim, start, end, step)
+  src: slice_scatter_symint(grad, zeros_like(self), dim, start, end, step)
+  result: auto_linear
```
Maybe add a small comment about why we need this formula manually

jbschlosser replied:
it's because it has to be CompositeExplicit so we can override it for a subclass

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)
    * Need to pass view_group info down into `gen_inplace_or_view.py` to get the proper inverse op name

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce slice_inverse() op"


Introduces a new op `slice_inverse()`. This is used in the reverse view_func for slice and several other ops (e.g. `split_with_sizes`, `chunk`). It's implemented behind the scenes by a call to `as_strided()`, but it's easier for subclasses to implement the more limited `slice_inverse()` than the full `as_strided()`. This PR:
* Introduces the op itself
* Updates all relevant functional inverses to call `slice_inverse()` instead of `as_strided()` directly
* Makes codegen changes to allow `slice_scatter()` to be the copy variant for `slice_inverse()`
    * Need to avoid view_copy codegen (assumes if view name ends in inverse, we don't need to gen one, which is possibly a bad assumption)

albanD / soulitzer / bdhirsh: I'm most interested in your thoughts on the codegen changes and whether this is the right way to go.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
